feat: add `delete_metadata_properties` and `_RemoteMetadataProperty.delete`

alvarobartt opened the pull request:
# Description

This PR adds both the method `delete_metadata_properties` shared by both the `FeedbackDataset` and the `RemoteFeedbackDataset`, and also the `pydantic.BaseModel` named `_RemoteMetadataProperty` that implements the method `delete` to be able to delete a `metadata_property` from a `RemoteFeedbackDataset` in-place i.e. a `FeedbackDataset` in Argilla.

## Code-snippet

* Delete `metadata_properties` locally
  ```python
  import argilla as rg
  
  dataset = rg.FeedbackDataset(
      fields=[...],
      questions=[...],
      metadata_properties=[
          TermsMetadataProperty(name="terms-metadata", values=["a", "b", "c"]),
          IntegerMetadataProperty(name="integer-metadata", min=0, max=10),
          FloatMetadataProperty(name="float-metadata", min=0, max=10),
      ],
  )
  
  deleted_metadata_properties = dataset.delete_metadata_properties(["terms-metadata", "integer-metadata", "float-metadata"])
  ```

* Delete `metadata_properties` remotely
  ```python
  import argilla as rg
  
  dataset = rg.FeedbackDataset(
      fields=[...],
      questions=[...],
      metadata_properties=[
          TermsMetadataProperty(name="terms-metadata", values=["a", "b", "c"]),
          IntegerMetadataProperty(name="integer-metadata", min=0, max=10),
          FloatMetadataProperty(name="float-metadata", min=0, max=10),
      ],
  )
  remote_dataset = dataset.push_to_argilla(name="my-dataset", workspace="my-workspace")
  
  # Alternative 1 -> via `delete_metadata_properties`
  deleted_metadata_properties = remote_dataset.delete_metadata_properties(["terms-metadata", "integer-metadata", "float-metadata"])
  # Alternative 2 -> via `delete` in-place
  for metadata_property in remote_dataset.metadata_properties:
    deleted_metadata_property = metadata_property.delete()
  ```

**Type of change**

- [X] New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

- [x] Add unit tests for `delete_metadata_properties` on `FeedbackDataset` (local only)
- [x] Add integration tests for `delete_metadata_properties` on `RemoteFeedbackDataset` (remote only)
- [x] Add integration tests for `_RemoteMetadataProperty.delete` via `RemoteTermsMetadataProperty`, `RemoteIntegerMetadataProperty`, and `RemoteFloatMetadataProperty`.
- [x] Add integration tests for `delete_metadata_property` in the Python SDK

**Checklist**

- [ ] I added relevant documentation
- [x] follows the style guidelines of this project
- [x] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [x] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

alvarobartt committed:
feat: `add_metadata_properties` in `RemoteFeedbackDataset`

alvarobartt committed:
fix: check `_metadata_properties is not None`

alvarobartt committed:
fix: check `metadata_properties` duplicated names

alvarobartt committed:
Merge branch 'feature/support-for-metadata-filtering-and-sorting' of github.com:argilla-io/argilla into feat/add-metadata-properties-to-existing

alvarobartt committed:
fix: return value after `validator`

alvarobartt committed:
docs: update `CHANGELOG.md`

alvarobartt committed:
feat: add `_validate_metadata_property`

alvarobartt committed:
feat: add `add_metadata_property`

alvarobartt committed:
fix: avoid circular `ArgillaMixin` imports

alvarobartt committed:
test(unit): `_validate_metadata_property`

alvarobartt committed:
Merge branch 'feature/support-for-metadata-filtering-and-sorting' into feat/add-metadata-properties-to-existing

alvarobartt committed:
Merge branch 'feature/support-for-metadata-filtering-and-sorting' into feat/add-metadata-properties-to-existing

alvarobartt committed:
fix: `_pydantic_field_with_validator` type-hint and default

alvarobartt committed:
fix: provide `parsed` instead of `httpx.Response`

alvarobartt committed:
test(integration): `add_metadata_property`

alvarobartt committed:
fix: use `name` in `_validate_metadata_property`

alvarobartt committed:
test(unit): `add_metadata_property{,_errors}`

alvarobartt committed:
fix(style): replace `super()` with `self`

Co-authored-by: Gabriel Martín Blázquez <gmartinbdev@gmail.com>

alvarobartt committed:
fix(docs): typo and missing docstring

Co-authored-by: Gabriel Martin <gabriel@argilla.io>

alvarobartt committed:
refactor: `_validate` to `_unique`

The method `_validate_metadata_property` is now `_unique_metadata_property`

alvarobartt committed:
fix: `_pydantic_field_with_validator` optional fields

alvarobartt committed:
fix(test): update `match` exception string

alvarobartt committed:
Merge branch 'feature/support-for-metadata-filtering-and-sorting' of github.com:argilla-io/argilla into feat/delete-metadata-properties

alvarobartt committed:
feat: add `delete_metadata_property` in SDK

alvarobartt committed:
fix: wrong `ValueError` exception message

alvarobartt committed:
feat: add `_RemoteMetadataProperty`, `id`, and `client`

alvarobartt committed:
fix: add `client` arg in `_parse_to_remote_metadata_property`

alvarobartt committed:
fix(type): endpoint name to `metadata-properties`

alvarobartt committed:
feat: `RemoteFeedbackDataset._metadata_properties_mapping` to be fetched from Argilla

alvarobartt committed:
feat: add `FeedbackDataset.delete_metadata_properties`

alvarobartt committed:
feat: fetch `metadata_properties` from Argilla

alvarobartt committed:
feat: add `RemoteFeedbackDataset.delete_metadata_properties`

alvarobartt committed:
Merge branch 'feature/support-for-metadata-filtering-and-sorting' of github.com:argilla-io/argilla into feat/delete-metadata-properties

alvarobartt committed:
Merge branch 'feature/support-for-metadata-filtering-and-sorting' of github.com:argilla-io/argilla into feat/delete-metadata-properties

frascuchon committed:
Merge branch 'feature/support-for-metadata-filtering-and-sorting' into feat/delete-metadata-properties

frascuchon commented on src/argilla/client/feedback/dataset/remote/dataset.py:
```diff
@@ -465,10 +466,58 @@ def add_metadata_property(
+        for metadata_property in metadata_properties:
+            try:
+                metadata_property = metadata_properties_api_v1.delete_metadata_property(
+                    client=self._client, id=self.id
```
```suggestion
                    client=self._client, id=metadata_property.id
```

frascuchon commented on src/argilla/client/feedback/dataset/remote/dataset.py:
```diff
@@ -465,10 +466,58 @@ def add_metadata_property(
+            try:
+                metadata_property = metadata_properties_api_v1.delete_metadata_property(
+                    client=self._client, id=self.id
+                ).parsed
```
or why not just
```suggestion
                metadata_property.delete()
```

alvarobartt replied:
True, I implemented first the `delete_metadata_property` method and then I forgot to change it, nice catch!

frascuchon commented on src/argilla/client/feedback/schemas/remote/metadata.py:
```diff
@@ -12,20 +12,41 @@
-class RemoteTermsMetadataProperty(TermsMetadataProperty, RemoteSchema):
+class _RemoteMetadataProperty(RemoteSchema):
+    @allowed_for_roles(roles=[UserRole.owner, UserRole.admin])
+    def delete(self) -> "AllowedMetadataPropertyTypes":
```
We can use this method in the bulk deletion, right?

frascuchon replied:
The problem with this approach is that again we're providing different ways to delete metadata properties.

Please, keep it simple, and align both classes as much as possible. I would say to use `delete_metadata_property` as a general way, simple to use and supported for both, local and remote datasets.

alvarobartt replied:
But then we should remove all the in-place methods defined for all the remote instances, IMO aligning both is a huge work and users don't use Argilla to manage datasets locally but within Argilla, so local support should probably be discontinued...

frascuchon commented on src/argilla/client/feedback/dataset/remote/dataset.py:
```diff
@@ -465,10 +466,58 @@ def add_metadata_property(
+        deleted_metadata_properties = []
+        for metadata_property in metadata_properties:
+            try:
+                metadata_property = metadata_properties_api_v1.delete_metadata_property(
```
We need to check first the property is part of the current dataset.

alvarobartt committed:
revert: `Extra.forbid` as `Extra.ignore` is unsafe

I believe this patch was done before due to an incompatibility right after https://github.com/argilla-io/argilla/pull/3784 which is already patched at https://github.com/argilla-io/argilla/pull/3829, so this is not required now, also because it's unsafe

alvarobartt committed:
fix(style): missing type-hints, wrong/unused imports and standard `warnings.warn`

alvarobartt committed:
refactor: mapping to `__question_name_to_id`

Moved from the `RemoteFeedbackRecords` class into the `RemoteFeedbackDataset`, removed `_question_id_to_name` as unused, and `_question_name_to_id` made private

alvarobartt committed:
fix: `datasets_api_v1.get_records` filtering args from `@property`

Use `@property` instead of defining those as a class method, and renamed those to end with `query_strings`

alvarobartt committed:
feat: add `ArgillaMetadataPropertiesMixin`

alvarobartt committed:
improve: remove `_metadata_properties_mapping`

alvarobartt committed:
test(integration): `delete` method in-place

alvarobartt committed:
fix: remove `__question_name_to_id` variable

alvarobartt committed:
test(integration): `delete_metadata_properties` and `delete`

alvarobartt committed:
test(integration): `test_delete_metadata_property`

alvarobartt committed:
fix(style): `test_not_implemented_methods`

alvarobartt committed:
fix: `delete_metadata_properties` to update `self`

alvarobartt committed:
test(unit): `test_delete_metadata_properties`

alvarobartt committed:
docs: update `CHANGELOG.md`

alvarobartt committed:
Merge branch 'feature/support-for-metadata-filtering-and-sorting' into feat/delete-metadata-properties
